src/lib/SourceRename.hs

```diff
@@ -284,16 +307,17 @@ sourceRenameUBinder :: (Distinct o, Renamer m, Color c)
                     -> m o a
 sourceRenameUBinder asUVar ubinder cont = case ubinder of
   UBindSource b -> do
-    SourceMap sourceMap <- askSourceMap
+    SourceMap local    <- askLocalSourceMap
+    SourceMap imported <- askImportedSourceMap
     mayShadow <- askMayShadow
-    unless (mayShadow || not (M.member b sourceMap)) $
+    let shadows = M.member b local || M.member b imported
+    when (shadows && not mayShadow) $
```

Comment:
nit: put the `mayShadow` condition first, to avoid evaluating `shadows` when it's not necessary